File: fence/test/fence/stats_test.exs
defmodule Fence.StatsTest do
  use Fence.DataCase, async: false

  import Fence.Factory

  alias Fence.{Geofences, Groups, Locations, Stats}

  describe "get_user_stats/1" do
    test "returns empty list when user has no home claimed" do
      user = create_user()
      _group = create_group(user)

      assert Stats.get_user_stats(user.id) == []
    end

    test "returns stats with home visit count" do
      user = create_user()
      group = create_group(user)
      geofence = create_geofence(group, user, %{"name" => "Home"})

      {:ok, _} = Geofences.claim_home(user.id, geofence.id, group.id)

      # Simulate 2 entries to home
      Locations.update_geofence_state(user.id, MapSet.new([geofence.id]), MapSet.new())
      Locations.update_geofence_state(user.id, MapSet.new(), MapSet.new([geofence.id]))
      Locations.update_geofence_state(user.id, MapSet.new([geofence.id]), MapSet.new())

      [stat] = Stats.get_user_stats(user.id)

      assert stat.group_id == group.id
      assert stat.group_name == "Test Group"
      assert stat.home_geofence_name == "Home"
      assert stat.home_visit_count == 2
      assert stat.housemates == []
      assert stat.your_top_geofences == []
    end

    test "returns user's top non-home geofences" do
      user = create_user()
      group = create_group(user)
      home = create_geofence(group, user, %{"name" => "Home"})
      work = create_geofence(group, user, %{"name" => "Work"})
      gym = create_geofence(group, user, %{"name" => "Gym"})

      {:ok, _} = Geofences.claim_home(user.id, home.id, group.id)

      # 3 visits to work
      for _ <- 1..3 do
        Locations.update_geofence_state(user.id, MapSet.new([work.id]), MapSet.new())
        Locations.update_geofence_state(user.id, MapSet.new(), MapSet.new([work.id]))
      end

      # 1 visit to gym
      Locations.update_geofence_state(user.id, MapSet.new([gym.id]), MapSet.new())
      Locations.update_geofence_state(user.id, MapSet.new(), MapSet.new([gym.id]))

      [stat] = Stats.get_user_stats(user.id)

      assert length(stat.your_top_geofences) == 2
      [first, second] = stat.your_top_geofences
      assert first.geofence_name == "Work"
      assert first.visit_count == 3
      assert second.geofence_name == "Gym"
      assert second.visit_count == 1
    end

    test "includes housemates with active visibility" do
      user = create_user(%{"display_name" => "Alice"})
      housemate = create_user(%{"display_name" => "Bob"})
      group = create_group(user)
      home = create_geofence(group, user, %{"name" => "Home"})

      # Add housemate to group
      {:ok, invite} = Groups.get_or_create_invite(group.id, user.id)
      {:ok, _} = Groups.join_by_invite_code(housemate.id, invite.code)
      {:ok, _} = Groups.share_visibility(user.id, group.id, housemate.id)

      # Both claim same home
      {:ok, _} = Geofences.claim_home(user.id, home.id, group.id)
      {:ok, _} = Geofences.claim_home(housemate.id, home.id, group.id)

      [stat] = Stats.get_user_stats(user.id)

      assert length(stat.housemates) == 1
      assert hd(stat.housemates).display_name == "Bob"
    end

    test "excludes housemates without active visibility" do
      user = create_user(%{"display_name" => "Alice"})
      stranger = create_user(%{"display_name" => "Charlie"})
      group = create_group(user)
      home = create_geofence(group, user, %{"name" => "Home"})

      # Add stranger to group, then revoke auto-shared visibility
      {:ok, invite} = Groups.get_or_create_invite(group.id, user.id)
      {:ok, _} = Groups.join_by_invite_code(stranger.id, invite.code)
      {:ok, _} = Groups.revoke_visibility(user.id, group.id, stranger.id)

      {:ok, _} = Geofences.claim_home(user.id, home.id, group.id)
      {:ok, _} = Geofences.claim_home(stranger.id, home.id, group.id)

      [stat] = Stats.get_user_stats(user.id)

      assert stat.housemates == []
    end
  end
end

File: mobile/test/models/stats_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:fence/models/stats.dart';

void main() {
  group('GeofenceVisitStat.fromJson', () {
    test('parses all fields correctly', () {
      final stat = GeofenceVisitStat.fromJson({
        'geofence_id': 'gf1',
        'geofence_name': 'Work',
        'visit_count': 15,
      });

      expect(stat.geofenceName, 'Work');
      expect(stat.visitCount, 15);
    });
  });

  group('HousemateStat.fromJson', () {
    test('parses with top geofences', () {
      final stat = HousemateStat.fromJson({
        'display_name': 'Alice',
        'top_geofences': [
          {'geofence_id': 'gf1', 'geofence_name': 'Office', 'visit_count': 10},
          {'geofence_id': 'gf2', 'geofence_name': 'Gym', 'visit_count': 5},
        ],
      });

      expect(stat.displayName, 'Alice');
      expect(stat.topGeofences, hasLength(2));
      expect(stat.topGeofences.first.geofenceName, 'Office');
      expect(stat.topGeofences.first.visitCount, 10);
    });

    test('parses with empty top geofences', () {
      final stat = HousemateStat.fromJson({
        'display_name': 'Bob',
        'top_geofences': <Map<String, dynamic>>[],
      });

      expect(stat.displayName, 'Bob');
      expect(stat.topGeofences, isEmpty);
    });
  });

  group('GroupStats.fromJson', () {
    test('parses full response', () {
      final stat = GroupStats.fromJson({
        'group_id': 'g1',
        'group_name': 'Family',
        'home_geofence_name': 'Home',
        'home_visit_count': 47,
        'housemates': [
          {
            'display_name': 'Alice',
            'top_geofences': [
              {'geofence_id': 'gf1', 'geofence_name': 'Work', 'visit_count': 15},
            ],
          },
        ],
        'your_top_geofences': [
          {'geofence_id': 'gf2', 'geofence_name': 'Office', 'visit_count': 22},
        ],
      });

      expect(stat.groupId, 'g1');
      expect(stat.groupName, 'Family');
      expect(stat.homeGeofenceName, 'Home');
      expect(stat.homeVisitCount, 47);
      expect(stat.housemates, hasLength(1));
      expect(stat.housemates.first.displayName, 'Alice');
      expect(stat.yourTopGeofences, hasLength(1));
      expect(stat.yourTopGeofences.first.geofenceName, 'Office');
    });

    test('parses with empty lists', () {
      final stat = GroupStats.fromJson({
        'group_id': 'g1',
        'group_name': 'Family',
        'home_geofence_name': 'Home',
        'home_visit_count': 0,
        'housemates': <Map<String, dynamic>>[],
        'your_top_geofences': <Map<String, dynamic>>[],
      });

      expect(stat.homeVisitCount, 0);
      expect(stat.housemates, isEmpty);
      expect(stat.yourTopGeofences, isEmpty);
    });
  });
}
